### PR TITLE
Configurable version and addons. Adding tests.

### PR DESCRIPTION
**Pull request by ktsakalozos**
Changes:
- List of addons to be enabled during the init phase
- Install MicroK8s from specific version
- Make unit tests passing
- Introduce GH actions

### CODE REVIEW

File: controllers/cloudinit/cloudinit_test.go [hunk 1]
```diff
@@ -20,250 +20,77 @@ import (
...
-		`-   path: /tmp/my-path
-    encoding: "base64"
-    content: |
-      aGk=`,
-		`-   path: /tmp/my-other-path
-    content: |
-      hi`,
+	expectedCommands := []string{
+		`snap install microk8s --classic --channel=1.24`,
+		`--token my_join_token`,
```
Comment (lferran) [inline controllers/cloudinit/cloudinit_test.go:66]: if it matters at all: should the token be specified in the `join` command instead? Just like below

File: controllers/cloudinit/controlplane_init.go [hunk 2]
```diff
@@ -149,12 +153,28 @@ func NewInitControlPlane(input *ControlPlaneInput) ([]byte, error) {
...
+	major, minor, err := extractVersionParts(input.Version)
+	if err != nil {
+		return nil, err
+	}
+	input.Version = generateSnapChannelArgument(major, minor)
+
+	// Get at least dns enabled
+	if input.Addons == nil {
+		input.Addons = []string{"dns"}
+	}
```
Comment (lferran) [inline controllers/cloudinit/controlplane_init.go:165]: is the dns addon one that "we always want there"? If so, maybe check if "dns" is in the list?

File: controllers/cloudinit/controlplane_init.go [hunk 3]
```diff
@@ -149,12 +153,28 @@ func NewInitControlPlane(input *ControlPlaneInput) ([]byte, error) {
 	input.WriteFiles = append(input.WriteFiles, input.AdditionalFiles...)
 	input.SentinelFileCommand = sentinelFileCommand
 	input.ControlPlaneEndpointType = "DNS"
+	major, minor, err := extractVersionParts(input.Version)
+	if err != nil {
+		return nil, err
+	}
+	input.Version = generateSnapChannelArgument(major, minor)
```
Comment (lferran) [inline controllers/cloudinit/controlplane_init.go:160]: why not simply allowing the user to work with the snap channel?
This way we can test `latest/edge` and not just `1.25`-like channels

File: controllers/microk8sconfig_controller.go [hunk 4]
```diff
@@ -397,6 +401,7 @@ func (r *MicroK8sConfigReconciler) handleJoiningWorkerNode(ctx context.Context,
 		JoinToken:        "abcdefgwijklmnopqrstuvwxyzABCDEF",
 		PortOfNodeToJoin: portOfNodeToConnectTo,
 		IPOfNodeToJoin:   ipOfNodeToConnectTo,
+		Version:          *machine.Spec.Version,
```
Comment (lferran) [inline controllers/microk8sconfig_controller.go:404]: One thing I don't get is why the version comes from the machine spec?

Shouldn't it be set only in the microk8s config CRD?

Comment (ktsakalozos) [reply controllers/microk8sconfig_controller.go:404]: I believe we cannot alter the machine spec [1] as it comes with the clusterapi. I wouldn't duplicate this field in our configs.

[1] https://github.com/kubernetes-sigs/cluster-api/blob/main/api/v1beta1/machine_types.go#L74

File: controllers/cloudinit/controlplane_init.go (hunk 3 above)
Comment (ktsakalozos) [reply controllers/cloudinit/controlplane_init.go:160]: This is a good point. I am going to create a task-item for us to add the channel as an option a user can set.

**Pull request merged**